File: go/reasoning/options.go
package reasoning

import "time"

// Options bound the reasoning adapter's behavior. All limits fail closed: an
// input that exceeds a bound is rejected with an error rather than being
// truncated into a partial tree (summaries are the one bounded field by
// design; they are truncated, never the parse aborted).
type Options struct {
	// MaxReasoningBytes caps the size of the extracted reasoning text.
	// Input whose reasoning exceeds this bound is rejected. Zero means the
	// default (64 KiB).
	MaxReasoningBytes int
	// MaxNodes caps the number of nodes a single tree may contain. Zero
	// means the default (64).
	MaxNodes int
	// MaxDepth caps the depth of the node chain. Zero means the default (32).
	MaxDepth int
	// MaxSummaryChars caps the length of every node summary. Zero means the
	// default (512).
	MaxSummaryChars int
	// Clock supplies timestamps for tree and nodes. Zero uses time.Now.
	Clock func() time.Time
}

// DefaultOptions returns the adapter limits used when no Options are given.
func DefaultOptions() Options {
	return Options{
		MaxReasoningBytes: 64 * 1024,
		MaxNodes:          64,
		MaxDepth:          32,
		MaxSummaryChars:   512,
	}
}

func (o Options) normalize() Options {
	d := DefaultOptions()
	if o.MaxReasoningBytes <= 0 {
		o.MaxReasoningBytes = d.MaxReasoningBytes
	}
	if o.MaxNodes <= 0 {
		o.MaxNodes = d.MaxNodes
	}
	if o.MaxDepth <= 0 {
		o.MaxDepth = d.MaxDepth
	}
	if o.MaxSummaryChars <= 0 {
		o.MaxSummaryChars = d.MaxSummaryChars
	}
	if o.Clock == nil {
		o.Clock = time.Now
	}
	return o
}

File: typescript/src/reasoning/options.ts
// Bounds for the reasoning adapter (mirrors go/reasoning/options.go).
//
// All limits fail closed: an input that exceeds a bound raises an error rather
// than being truncated into a partial tree (summaries are the one bounded
// field by design; they are truncated, never the parse aborted).

export interface ReasoningOptions {
  /** Cap on the size of the extracted reasoning text, in bytes. */
  max_reasoning_bytes?: number;
  /** Cap on the number of nodes a single tree may contain. */
  max_nodes?: number;
  /** Cap on the depth of the node chain. */
  max_depth?: number;
  /** Cap on the length of every node summary, in code points. */
  max_summary_chars?: number;
  /** Clock supplying timestamps for tree and nodes. Defaults to Date.now. */
  clock?: () => Date;
}

export const DEFAULT_MAX_REASONING_BYTES = 64 * 1024;
export const DEFAULT_MAX_NODES = 64;
export const DEFAULT_MAX_DEPTH = 32;
export const DEFAULT_MAX_SUMMARY_CHARS = 512;

export function normalizeOptions(opts?: ReasoningOptions): Required<ReasoningOptions> {
  return {
    max_reasoning_bytes: opts?.max_reasoning_bytes ?? DEFAULT_MAX_REASONING_BYTES,
    max_nodes: opts?.max_nodes ?? DEFAULT_MAX_NODES,
    max_depth: opts?.max_depth ?? DEFAULT_MAX_DEPTH,
    max_summary_chars: opts?.max_summary_chars ?? DEFAULT_MAX_SUMMARY_CHARS,
    clock: opts?.clock ?? (() => new Date()),
  };
}
